src/features/FreeRooms/FreeRooms.ts

```diff
@@ -0,0 +1,335 @@
+    }
+
+    //Each row is subdivided into 56 different segments
+    //with the first segment representing the time 8:00 AM
+    //and the last segment representing 9:45 PM.
+
+    //There are a additional 3 segments that are on the table but unaccounted for.
+
+    if (time.getHours() < 8 || time.getHours() > 22) {
+        //These times are outside building operation hours
+        return false;
+    }
+    let colspanTime = convertTimetoColspan(time);
+    let totalColspan = 0;
+    let flag = false;
+    let dataElements = row.querySelectorAll('td');
+    for (let td of dataElements) {
+        if (td.hasAttribute('colspan')) {
+            totalColspan += parseInt(td.getAttribute('colspan'));
+            if (totalColspan >= colspanTime) {
+                if (td.innerHTML === '') {
+                    flag = true;
+                } else {
+                    break;
+                }
+            }
+        }
+    }
+    return flag;
+}
+
+function convertTimetoColspan(time) {
+    /*
+    Input is a time given by a Date() object between
+    8 AM and 9:45 PM.
+    Output is the column of the table this time occupies on the
+    room scheduling website.
+    */
+
+    //8 AM -> 0, 9:45 PM -> 55
+
+    let hours = time.getHours();
+    let minutes = time.getMinutes();
+    return (hours - 8) * 4 + Math.floor(minutes / 15);
+}
+
+function getSemesterFromDate(date) {
+    /*
+    Input is a Date() object.
+
+    Output is the current semester that corresponds to that date.
+    */
+
+    //TODO: more accurately determine the current semester
+    //without relying only on the month
+
+    let month = date.getMonth();
+    if (month === 0) {
+        return 'W';
+    } else if (month >= 1 && month <= 4) {
+        return 'S';
+    } else if (month === 5) {
+        return 'A';
+    } else if (month >= 6 && month <= 7) {
+        return 'B';
+    }
+    return 'F';
+}
+
+async function getAvailableRooms(date) {
+    /*
+    Input is a Date() object.
+
+    Output is a Promise that on success
+    returns an array of rooms with no classes scheduled according
+    to the room scheduling website.
+    */
+
+    let availableRooms = [];
+
+    let year = date.getFullYear();
+    let semester = getSemesterFromDate(date);
+    let day = date.getDay();
+
+    let tables = await getTablesPerRoom(year, semester);
+    for (let room in tables) {
+        let row = getDaysRow(day, tables[room]);
+        if (hasNoScheduledEvent(row, date)) {
+            availableRooms.push(room);
+        }
+    }
+    return availableRooms;
+}
+
+chrome.runtime.onInstalled.addListener(details => {
+    getAvailableRooms(new Date()).then(rooms => {
+        chrome.storage.local.set({availableRooms: rooms});
+        chrome.alarms.create('rooms', {
+            periodInMinutes: 15,
+        });
+    });
+
+    chrome.alarms.onAlarm.addListener(alarm => {
+        if (alarm.name === 'rooms') {
+            getAvailableRooms(new Date()).then(rooms => {
```

Comment:
Consider using [moment.js](https://momentjs.com/) for anything related to dates and times. It integrates very easily with Javascript Date objects and it's already part of the project dependencies.